### PR TITLE
Add IsNullFixer

### PR DESCRIPTION
**Pull request by kalessil**
- [x] Re-create the fixer
- [x] Fix bugs discovered in #1304 (assignments, ternaries)
- [x] Add option for yoda/no-yoda fixing style

### CODE REVIEW

**Comment by kalessil** [comment]
Ready for a review @keradus //cc @SpacePossum

**Comment by keradus** [comment]
out of the curiosity: self-checking the Fixer class by utests were nice for you or not really ?

File: src/Fixer/LanguageConstruct/IsNullFixer.php [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                }
+
+                // get rid of the root namespace when it used and check if the inversion operator provided
+                $tokens->removeTrailingWhitespace($prevTokenIndex);
+                $tokens[$prevTokenIndex]->clear();
+            }
+
+            // check if inversion being used, text comparison is due to not existing constant
+            $isInvertedNullCheck = false;
+            if ('!' === $tokens[$inversionCandidateIndex]->getContent()) {
```
Comment (SpacePossum) [inline src/Fixer/LanguageConstruct/IsNullFixer.php]:
```php
if ($tokens[$inversionCandidateIndex]->equals('!')) {
```

File: src/Fixer/LanguageConstruct/IsNullFixer.php [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+                $tokens[$matches[1]]->clear();
+            }
+
+            // sequence which we'll use as a replacement
+            $replacement = array(
+                new Token(array(T_STRING, 'null')),
+                new Token(array(T_WHITESPACE, ' ')),
+                new Token($isInvertedNullCheck ? array(T_IS_NOT_IDENTICAL, '!==') : array(T_IS_IDENTICAL, '===')),
+                new Token(array(T_WHITESPACE, ' ')),
+            );
```
Comment (SpacePossum) [inline src/Fixer/LanguageConstruct/IsNullFixer.php]: missing empty line

File: src/Fixer/LanguageConstruct/IsNullFixer.php [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    {
+        return $tokens->isTokenKindFound(T_STRING);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function fix(\SplFileInfo $file, Tokens $tokens)
+    {
+        $sequenceNeeded = array(array(T_STRING, 'is_null'), '(');
```
Comment (SpacePossum) [inline src/Fixer/LanguageConstruct/IsNullFixer.php]: can be `static`

File: tests/Fixer/LanguageConstruct/IsNullFixerTest.php [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+        $configuration = static::getObjectAttribute($fixer, 'configuration');
+        static::assertFalse($configuration['use_yoda_style']);
+    }
+
+    /**
+     * @dataProvider provideExamples
+     *
+     * @param mixed      $expected
+     * @param null|mixed $input
```
Comment (SpacePossum) [inline tests/Fixer/LanguageConstruct/IsNullFixerTest.php]: `mixed` -> `string`

**Comment by SpacePossum** [comment]
Looking good :)
Please add the following test cases (these pass btw, so :+1:)
```php
// config: not yoda style, nested calls
        $this->doTest(
            '<?php $b = a(a(a(b() === null) === null) === null) === null;',
            '<?php $b = \is_null(a(\is_null(a(\is_null(a(\is_null(b())))))));'
        );
```

```php
// config: yoda style, nested calls
            array(
                '<?php null === a(null === a(null === a(null === b())));',
                '<?php \is_null(a(\is_null(a(\is_null(a(\is_null(b())))))));',
            ),
```

File: src/Fixer/LanguageConstruct/IsNullFixer.php (hunk 2 above)
Comment (kalessil) [inline src/Fixer/LanguageConstruct/IsNullFixer.php]: Empty line was added, but GH didn't recognize it.

**Comment by kalessil** [comment]
Added the suggested test cases as well.

File: src/Fixer/LanguageConstruct/IsNullFixer.php [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition()
+    {
+        return new FixerDefinition(
+            'Replaces is_null(parameter) expression with `null === parameter`.',
+            array(new CodeSample("<?php\n\$a = is_null(\$b);")),
```
Comment (keradus) [inline src/Fixer/LanguageConstruct/IsNullFixer.php]: please add sample with non-default configuration

**Comment by kalessil** [comment]
Added

**Comment by SpacePossum** [comment]
awesome :)
:+1:

File: src/Fixer/LanguageConstruct/IsNullFixer.php [hunk 6]
```diff
@@ -172,7 +172,10 @@ public function getDefinition()
...
         return new FixerDefinition(
             'Replaces is_null(parameter) expression with `null === parameter`.',
-            array(new CodeSample("<?php\n\$a = is_null(\$b);")),
+            array(
+                new CodeSample("<?php\n\$a = is_null(\$b);"),
+                new CodeSample("<?php\n\$a = is_null(\$b);", array('use_yoda_style' => false)),
+            ),
             null,
             'The following can be configured: `use_yoda_style => boolean`',
             array('use_yoda_style' => true),
```
Comment (keradus) [inline src/Fixer/LanguageConstruct/IsNullFixer.php]: self::$defaultConfiguration

**Comment by kalessil** [comment]
Corrected

File: src/Fixer/LanguageConstruct/IsNullFixer.php [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+                if ($isContainingDangerousConstructs) {
+                    array_unshift($replacement, new Token(array(')')));
+                }
+
+                $tokens[$isNullIndex]->clear();
+                $tokens->removeTrailingWhitespace($referenceEnd);
+                $tokens->overrideRange($referenceEnd, $referenceEnd, $replacement);
+            }
+
+            // nested is_null's support
```
Comment (GrahamCampbell) [inline src/Fixer/LanguageConstruct/IsNullFixer.php]: No apostrophe here **(I think???)**. This is a very common mistake in English grammar. The support does not **belong to** is_null, you're just using the **plural form** of is_null, right?

File: src/Fixer/LanguageConstruct/IsNullFixer.php [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+            $inversionCandidateIndex = $prevTokenIndex = $tokens->getPrevMeaningfulToken($matches[0]);
+            $prevToken = $tokens[$prevTokenIndex];
+            if ($prevToken->isGivenKind(array(T_DOUBLE_COLON, T_NEW, T_OBJECT_OPERATOR, T_FUNCTION))) {
+                continue;
+            }
+
+            // handle function references with namespaces
+            if ($prevToken->isGivenKind(T_NS_SEPARATOR)) {
+                $inversionCandidateIndex = $twicePrevTokenIndex = $tokens->getPrevMeaningfulToken($prevTokenIndex);
+                /* @var Token $twicePrevToken */
```
Comment (GrahamCampbell) [inline src/Fixer/LanguageConstruct/IsNullFixer.php]: oops, this is not actually phpdoc :P

File: src/Fixer/LanguageConstruct/IsNullFixer.php (hunk 7 above)
Comment (kalessil) [inline src/Fixer/LanguageConstruct/IsNullFixer.php]: @GrahamCampbell : I'll correct this, thanks for spotting this out. Btw, would you also please check other comments in the file for English correctness?

Comment (GrahamCampbell) [inline src/Fixer/LanguageConstruct/IsNullFixer.php]: tbh, i think it reads best without the "s" at all :)

**Comment by kalessil** [comment]
Something wrong with travis builds:

```log
1) PhpCsFixer\Tests\ReadmeTest::testIfReadmeFileIsCorrect
README.rst file is not up to date! Do not modify it manually! Regenerate readme with command: `php php-cs-fixer readme > README.rst`.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
 * **is_null**
-   | Replaces is_null(parameter) expression with ``null === parameter``.
+   | Replaces is_null(parameter) expression with `null === parameter`.
    | *Rule is: configurable, risky.*
```

**Comment by SpacePossum** [comment]
I think you need to regen. the README (maybe rebase first), the generator has been modified not long ago

**Comment by keradus** [comment]
Like the utest message ask you for :D

**Comment by kalessil** [comment]
Rebased, force-pushed. Will reopen the PR as it green again.

**Comment by SpacePossum** [comment]
it won't get green here if you don't reopen ;)
did you do the regen?
```bash
$ php php-cs-fixer readme > README.rst` 
```

**Comment by kalessil** [comment]
Did exactly the same.

If you login into travis with github account, it runs automated build for commits automatically. So I actually nedd just to wait for several minutes and will re-open the PR.

**Comment by SpacePossum** [comment]
ah awesome : )

File: src/Fixer/LanguageConstruct/IsNullFixer.php [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    private static $configurableOptions = array('use_yoda_style');
+    private static $defaultConfiguration = array('use_yoda_style' => true);
+
+    /**
+     * @var array
+     */
+    private $configuration;
+
+    /**
+     * 'use_yoda_style' can be configured wit a boolean value.
```
Comment (keradus) [inline src/Fixer/LanguageConstruct/IsNullFixer.php]: typo wit**h**

**Comment by kalessil** [comment]
Are we fine with the PR? Perhaps someone can even merge it?

**Comment by SpacePossum** [comment]
please fix the little typo
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2415#discussion_r94065279

and maybe update the config hint to:
```php
    /**
      * @var array<string, bool>
      */
     private $configuration;
```

However IMO these are not blocking for merging. I'm :+1:

File: src/Fixer/LanguageConstruct/IsNullFixer.php (hunk 9 above)
Comment (kalessil) [inline src/Fixer/LanguageConstruct/IsNullFixer.php]: Oh, I missed this. Fixed now.

**Comment by keradus** [comment]
Thank you @kalessil.

**Comment by keradus** [comment]
great work here @kalessil ;)

**Comment by SpacePossum** [comment]
thanks @kalessil, awesome new fixer :D

**Comment by kalessil** [comment]
Thank you @keradus , @SpacePossum ;) It's my pleasure to contribute here )

**Pull request closed without merging**